# Feedback from previous weeks and other and hints

1. Be careful about your environment remembering variables. Make sure your code works in a new _clean_ environment. In Colab: `Runtime`->`restart Runtime`, in Anaconda's Jupyter: `Kernel`->`Restart`.
2. Graphs without labels (or units when appropriate) are not worth any point.
3. Do put in sufficient explanatory comments in your code.
4. Functions are very important. Do look up the video on the Safari O'Reilly ressource if you are still not clear on them !

For this week you can use these imports at the start of your programs:

In [ ]:
import numpy as np
import matplotlib.pyplot as plt

We will use a new module `uncertainties`, which is **not** standard in the colab environment. You will have to first run:

In [15]:
! pip install -q uncertainties

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.1/60.1 kB 2.2 MB/s eta 0:00:00


To install the module before you can import it:

In [16]:
import uncertainties as uc
import uncertainties.umath as um # for maths functions

# Introduction
In the practical classes PX2133/PX2233 and PX2338 (Obs tech), as well as your year 3/4 project, a lot of emphasis is placed on the determination and mathematical handling of errors.
The uncertainties module allows us to deal very easily with [error propagation](https://en.wikipedia.org/wiki/Propagation_of_uncertainty). For this sheet you should remind yourself about error bars in measurements and about propagation of uncertainties. Take an example from your lab handbook:

**Example 1**: If the length of a rectangle is $1.24\pm0.02 m$ and its breadth is $0.61\pm0.01 m$, what is its area and the error in the area? The following code snippet solves this problem in a few lines.

In [ ]:
L = uc.ufloat(1.24, 0.02)
W = uc.ufloat(0.61, 0.01)
print ('Area is:', L*W, 'm^2') # Do remember to add the units when printing!

Area is: 0.756+/-0.017 m^2


**Note**: For the area itself, it's fairly straightforward:

In [ ]:
1.24*0.61

0.7564

However, for the error bar on this number:

In [ ]:
0.02*0.01

0.0002

does not work. Instead, the [error progation formula](https://en.wikipedia.org/wiki/Propagation_of_uncertainty#Example_formulae) gives:

In [ ]:
np.abs(1.24*0.61)*np.sqrt((0.02/1.24)**2+(0.01/0.61)**2)

0.01739540169125163

So the area is $0.756\pm0.017 m^2$. `uncertainties` obviously saves a lot of work, even for such a simple case. You can also take a look at the web site uncertainties hosted at https://pythonhosted.org/uncertainties/user_guide.html. In particular, [this section](https://pythonhosted.org/uncertainties/user_guide.html#access-to-the-uncertainty-and-to-the-nominal-value) shows some of the properties of a `ufloat` you can access directly.

**Example 2**: A reference object is $10.0\pm0.0001 m$ long, and makes a viewing angle of $0.62\pm0.02 rad$. How far is it?

In [ ]:
L = uc.ufloat(10.0, 0.0001)
theta = uc.ufloat(0.62,0.02)

Distance = (L/2)/um.tan(theta/2)

print ('Distance is:', Distance.nominal_value, 'm, with an error of:', Distance.std_dev)

Distance is: 15.609024890896208 m, with an error of: 0.537283338762715


Note the need to use "umath" functions (like `um.tan()` instead of `np.tan()`), and how to get the nominal value and the standard deviation of the uncertainties objects. To get nicer looking output, such as controlling the number of significant digits printed, you can use the information about formatting at https://docs.python.org/3/tutorial/inputoutput.html. In the exercises below you need to print the values to the screen. (Don’t forget units.)

# Exercises
This must be marked before you leave the lab. Mark weighting is in brackets.
**Save your work to GitHub after having run all cells with `Runtime` -> `Restart and run all`. And do not change the notebook's filename.** Do add comments to your code, you'll lose points if your code is hard to understand. Graphs without labels (or units when appropriate) are not worth any point.

## Exercise 0
[0] With some approximations, we have measured the mass of the following black-holes:
```
"35.6+/-3.9","30.6+/-3.7","63.1+/-3.2","23.2+/-9.8","13.6+/-4.5","35.7+/-6.8","13.7+/-6.0","7.7+/-2.4","20.5+/-4.0"
```
Compute for each (with error-bars) their lifetime due to Hawking radiation:
$$
t = \left(\frac{M}{M_{\odot}}\right)^3\,\times\,2.097\,\times\,10^{67} yr
$$

(this exercise is for demonstration purposes and won't be marked)

## Exercise 1
[2] An object is measured to travel a distance $x = 5.1 \pm 0.4 m$ during a time of $t = 0.4 \pm 0.1 s$. What is the average velocity and the error in the average velocity?

In [ ]:
# v = x/t
#define each of the variables
x = uc.ufloat(5.1, 0.4)
t = uc.ufloat(0.4, 0.1)
#perform the calculation
print(f"the area is: {x/t} ms^-1")

the area is: 12.7+/-3.3 ms^-1


## Exercise 2
[2] An enterprising cow attempts to jump over the moon by jumping vertically into the air with initial speed $v_0=4.0\pm0.2 m/s$. After a time $t=0.60\pm0.06s$, the height of the cow is $h = v_0t-\frac{1}{2}g t^2 = 0.636 m$. What is the uncertainty in $h$? Take $g$ as exactly $9.81 ms^{-2}$.

In [ ]:
# why is there an enterprising cow?????
#define the variables
v_0 = uc.ufloat(4.0, 0.2)
t = uc.ufloat(0.6, 0.06)
g = 9.81
#perform the calculation
h = v_0*t - 0.5*g*t**2

print(f"the height the cow jumps is {h.nominal_value} m and the uncertainty is {h.std_dev}")
print(f"the uncertainty in the height is {h.std_dev}")

the height the cow jumps is 0.6341999999999999 m and the uncertainty is 0.1649399454346945
the uncertainty in the height is 0.1649399454346945


## Exercise 3
[2] In an optics experiment the object distance $u$ is measured to be 20cm and the image distance $v$ is 10cm, both to an accuracy of 0.5cm. Find the focal length $f$ of the lens using the formula:

$$ \frac{1}{u}+\frac{1}{v}=\frac{1}{f}$$

In [ ]:
#define the variables
u = uc.ufloat(20, 0.5)
v = uc.ufloat(10, 0.5)

#perform the calculation
f = 1/(1/u + 1/v)

print(f"the focal length is {f.nominal_value} cm")

the focal length is 6.666666666666666 cm


## Exercise 4
[2] Two students each measure the refractive index of water. Jack measures a value of $1.33 \pm 0.03$ while Jill measures $1.28 \pm 0.02$. Are these values in agreement? *You do have to think a bit about this one...*

In [ ]:
#to see if the values aggree, you can look at the range of refractive indexes both measured
#come back to this question
print("*******************????????????")
print(f"for jack the max refractive index is 1.36 and the min refractive index is 1.30")
print(f"for jill the max refractive index is 1.30 and the min refractive index is 1.26")
print(f"Jack and Jill only have one value of the refractive index that they can agree on: 1.30, so they can agree on only that value")

*******************????????????
for jack the max refractive index is 1.36 and the min refractive index is 1.30
for jill the max refractive index is 1.30 and the min refractive index is 1.26
Jack and Jill only have one value of the refractive index that they can agree on: 1.30, so they can agree on only that value


## Exercise 5
[2] The damped resonance frequency $\omega_{res}$ of an oscillating system is related to the (un-damped) natural angular frequency $\omega_0$ and the damping coefficient $\alpha$ by:

$$\omega_{res} = \sqrt{ \omega_0^2 - 2\alpha^2}$$

Find $f_0$ if the measured resonance frequency $f_{res}$ is $23.2\pm0.1 Hz$ and the measured damping coefficient is $19.5\pm0.5s^{-1}$.


In [ ]:
#define the variables
f_res = uc.ufloat(23.3, 0.1)
a = uc.ufloat(19.5, 0.5)
#sustitute into w = 2*pi*f into both w_res and w_0
#then rearrange the equation above for f_0
# use um.sqrt instead of np.sqrt, because np.__ will not take account of the uncertainties
f_0 = (um.sqrt((2*np.pi*f_res)**2 + (2*(a**2))))/2*np.pi

#perform the calculation
print(f"the value of f_0 is {f_0.nominal_value} Hz, and the uncertainty is {f_0.std_dev}")

the value of f_0 is 234.0061536973049 Hz, and the uncertainty is 0.9914570714492409


## Exercise 6
[2] Suppose you have the following equation from one of your lab experiments:

$$f=\frac{c}{2}\sqrt{\frac{n_x^2}{L_x^2}+\frac{n_y^2}{L_y^2}+\frac{n_z^2}{L_z^2}}$$

where $f$ is the resonant frequency of sound waves in a box of sides $L_x$, $L_y$ and $L_z$ in length and the $n_x$ etc. are integers. $L_x = 10.2\pm0.2m$, $L_y = 5.2\pm0.3m$ and $L_z = 20.0\pm0.1 m$, while $c = 331.3 + T * 0.606 \,m\,s^{-1}$ is the temperature-dependent speed of sound, and the temperature $T$ is $23 \pm 1^\circ C$.
Calculate $f$ and the error in $f$ for the following values of $(nx,ny,nz)= (1,1,1), (1,1,2)$ and $(2,1,1)$.


In [ ]:
#define the variables
#part 1
L_x = uc.ufloat(10.2, 0.2)
L_y = uc.ufloat(5.2, 0.3)
L_z = uc.ufloat(20, 0.1)
T = uc.ufloat(23,1)
c = 331.3 + T*0.606
n_1 = np.array([1, 1, 1])
n_2 = np.array([1, 1, 2])
n_3 = np.array([2, 1, 1])

#the first formula
f_1 = c/2*(um.sqrt(((n_1[0])**2/(L_x)**2) + ((n_1[1])**2/(L_y)**2) + ((n_1[2])**2/(L_z)**2)))

#the second formula
f_2 = c/2*(um.sqrt(((n_2[0])**2/(L_x)**2) + ((n_2[1])**2/(L_y)**2) + ((n_2[2])**2/(L_z)**2)))

#the third formula
f_3 = c/2*(um.sqrt(((n_3[0])**2/(L_x)**2) + ((n_3[1])**2/(L_y)**2) + ((n_3[2])**2/(L_z)**2)))


#displaying the answer
print(f"part 1")
print(f"the value of the resonant frequncy when (nx,ny,nz) = (1,1,1) is {f_1.nominal_value} and the uncertainty is {f_1.std_dev}")
print("\n")
print(f"part 2")
print(f"the value of the resonant frequncy when (nx,ny,nz) = (1,1,2) is {f_2.nominal_value} and the uncertainty is {f_2.std_dev}")
print("\n")
print(f"part 3")
print(f"the value of the resonant frequncy when (nx,ny,nz) = (2,1,1) is {f_3.nominal_value} and the uncertainty is {f_3.std_dev}")

part 1
the value of the resonant frequncy when (nx,ny,nz) = (1,1,1) is 38.24745337649154 and the uncertainty is 1.6700605379580133


part 2
the value of the resonant frequncy when (nx,ny,nz) = (1,1,2) is 41.06516265028539 and the uncertainty is 1.5562781935426817


part 3
the value of the resonant frequncy when (nx,ny,nz) = (2,1,1) is 48.18791630742113 and the uncertainty is 1.4018271041233974


## Exercise 7
[4] The reflection coefficient $R_\parallel$ for parallel plane-polarised light reflected from a surface is given by the equation:

$$ R_\parallel = \frac{\tan^2(\theta_i - \theta_t)}{\tan^2(\theta_i + \theta_t)} $$

Calculate the error in $R_\parallel$ given measurements $\theta_i = (78 \pm 1)^\circ$ and $\theta_t = (40 \pm 1)^\circ$.

In [ ]:
#define the variable
theta_i = uc.ufloat(78, 1)
theta_t = uc.ufloat(40, 1)

#perform the calculation
R_parallel = (um.tan(theta_i - theta_t))**2/(um.tan(theta_i + theta_t))**2

#display the answer
print(f"the reflection coefficient is {R_parallel.nominal_value} and the uncertainty is {R_parallel.std_dev}")
print(f"the uncertainty in the reflection coefficient is {R_parallel.std_dev}")

the reflection coefficient is 0.003572137026589645 and the uncertainty is 0.06506774439998199
the uncertainty in the reflection coefficient is 0.06506774439998199


## Exercise 8
[4] Calculate and print to the screen the fractional uncertainty, as a percentage to one
significant figure, of the fluid flow discharge coefficient $C_d$ from the equation

$$
C_d = \frac{\dot{m}\sqrt{1-\left(\frac{d}{D}\right)^4}}{Kd^2F\sqrt{\rho\Delta P}}
$$

where

\begin{align*}
    C_d &= \text{discharge coefficient}&& \text{(no units)} \\
    \dot{m} &= \text{mass flow rate}&& = 0.13 \pm 0.01kg\,s^{-1} \\
    d &= \text{orifice diameter}&& = 11\pm 1 mm \\
    D &= \text{pipe diameter}&& = 71 \pm 1 mm \\
    \rho &= \text{fluid density}&& =1.01\pm0.01g\,cm^{-3} \\
    \Delta P &= \text{differential pressure}&& =156 \pm 7 Pa \\
    K &= \text{a constant parameter}&& =\text{constant (no units)} \\
    F &= \text{thermal expansion factor}&& =\text{constant (no units)}
\end{align*}


In [ ]:
# defining the variables
m = uc.ufloat(0.13, 0.01)
d = uc.ufloat(11, 1)
D = uc.ufloat(71, 1)
rho = uc.ufloat(1.01, 0.01)
P = uc.ufloat(156, 7)
K = 1
F = 2
#perform the calculation
C_d = (m*um.sqrt(1-(d/D)**4))/(K*(d**2)*F*(um.sqrt(rho*P)))

#display the answer
print(f"the value of the flow discharge coefficient, C_d, is {C_d.nominal_value} and the error is {C_d.std_dev}")



the value of the flow discharge coefficient, C_d, is 4.278384914858706e-05 and the error is 8.507535196525102e-06


## Exercise 9: Optional problem (not marked)
If you have time and want to try something interesting, do the following problem by plotting in 2D:
 - Draw an equilateral triangle with vertices and coordinates: vertex 1: $(p_1,q_1)$; vertex 2: $(p_2, q_2)$; vertex 3: $(p_3, q_3)$.
 - Place a dot at an arbitrary point $P = (x_0, y_0)$ within this triangle.
 - Find the next point by selecting randomly an integer $n = 1 , 2, $  or $3$ :
    1. If 1 , place a dot halfway between P and vertex 1.
    2. If 2 , place a dot halfway between P and vertex 2.
    3. If 3 , place a dot halfway between P and vertex 3.
 - Repeat the last two steps using the last dot as the new P.

Mathematically, the coordinates of successive points are given by the formulae

$$(x_{i+1},y_{i+1})=0.5[(x_i,y_i)+(p_n,q_n)]$$

and

$$n=int(1+3r_i),$$

where $r_i$ is a random number between 0 and 1 and where the $int()$ function outputs the closest integer smaller than or equal to the argument.

Try extending this to four vertices.